### PR TITLE
State and variables

### PR DESCRIPTION
**Pull request by aliabid94**
Created new component for storing state called a Variable. There can be multiple variables, all stored in the backend. Original state functionality in interfaces still supported through the "state" string shortcut. Can be tested through chatbot demo or simple_squares demo. Also created flashcards app demo, though the broken dataframe component makes the app not functional.

### CODE REVIEW

**Comment by FarukOzderim** [comment]
Questions: 
1. What is the purpose of `variable`?
2. Will backend just send updates of the variables or send them every time?

File: gradio.egg-info/requires.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
-aiohttp
 analytics-python
+aiohttp
```
Comment (FarukOzderim) [inline gradio.egg-info/requires.txt:2]: What are these unnecessary changes for?

File: gradio/blocks.py [hunk 2]
```diff
@@ -208,29 +208,46 @@ def render(self):
         self._id = Context.id
         Context.id += 1
 
-    def process_api(self, data: Dict[str, Any], username: str = None) -> Dict[str, Any]:
+    def process_api(
+        self,
+        data: Dict[str, Any],
+        username: str = None,
+        state: Optional[Dict[int, any]] = None,
```
Comment (FarukOzderim) [inline gradio/blocks.py:215]: let's create a docstring and explain state inside it?

File: gradio/blocks.py [hunk 3]
```diff
@@ -208,29 +208,46 @@ def render(self):
...
-                if predictions[i] is not None
-                else None
-                for i, output_id in enumerate(dependency["outputs"])
-            ]
+            output = []
+            for i, output_id in enumerate(dependency["outputs"]):
+                block = self.blocks[output_id]
+                if getattr(block, "stateful", False):
+                    state[output_id] = predictions[i]
+                    output.append(None)
```
Comment (FarukOzderim) [inline gradio/blocks.py:242]: why not append it to output?

File: gradio/blocks.py [hunk 4]
```diff
@@ -208,29 +208,46 @@ def render(self):
...
+                else:
+                    output.append(
+                        block.postprocess(predictions[i])
+                        if predictions[i] is not None
+                        else None
+                    )
+            else:
+                output = predictions
```
Comment (FarukOzderim) [inline gradio/blocks.py:250]: why do we override `output` after the loop is finished?

File: gradio/interface.py [hunk 5]
```diff
@@ -178,27 +179,19 @@ def __init__(
         if not isinstance(outputs, list):
             outputs = [outputs]
 
+        if "state" in inputs or "state" in outputs:
+            if len([i for i in inputs if i == "state"]) != 1 or len([o for o in outputs if o == "state"]) != 1:
```
Comment (FarukOzderim) [inline gradio/interface.py:183]: shall we break this check for readibility?

Comment (FarukOzderim) [inline gradio/interface.py:183]:
```suggestion
            state_input_count = len([i for i in inputs if i == "state"])
            state_output_count = len([o for o in outputs if o == "state"])
            if state_input_count  != 1 or state_output_count  != 1:
```

**Review by FarukOzderim — CHANGES_REQUESTED** [review]
Good work!

**Comment by aliabid94** [comment]
> 



> Questions:
> 
> 1. What is the purpose of `variable` and this PR? Was there no multiple state component support previously?
> 2. Will backend just send updates of the `variables `or send `variable `data every time?

1. There was a State component, but it was only possible to have one State component with an interface. It did not work with Blocks at all. Variables have replaced State.
2. Variable data is stored completely in the backend, unlike all other components. It is never passed to the frontend. 

The main purposes of variable are:
- Allow users to be able to store state between interactions without having them displayed in a frontend component
- Be able to store large binary data, e.g. model state, between interactions that may not be JSON serializable easily
- Bring statefulness to Blocks

File: gradio/blocks.py (hunk 3 above)
Comment (aliabid94) [reply gradio/blocks.py:242]: We don't want to send the state to the frontend. Variables are not displayed in the frontend, and may be massive or binary data. We have to append to preprocess because it is needed to run the function, but variable data is never sent to the frontend.

File: gradio/blocks.py (hunk 4 above)
Comment (aliabid94) [reply gradio/blocks.py:250]: Whoops typo - the return of this function should be output. Output is not overridden after the loop, its the 'else' clause where there is no postprocesssing, so the output is just the predictions.

File: gradio/interface.py (hunk 5 above)
Comment (aliabid94) [reply gradio/interface.py:183]: Done.

File: demo/flashcards/run.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
+import gradio as gr
```
Comment (abidlabs) [inline demo/flashcards/run.py:1]: Suggest naming this folder `blocks_flashcards`

File: demo/simple_squares/run.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import gradio as gr
```
Comment (abidlabs) [inline demo/simple_squares/run.py:1]: Suggest naming this folder `blocks_simple_squares` to be consistent with naming of all of the blocks demos.

File: gradio/interface.py [hunk 8]
```diff
@@ -178,27 +179,21 @@ def __init__(
         if not isinstance(outputs, list):
             outputs = [outputs]
 
+        if "state" in inputs or "state" in outputs:
```
Comment (abidlabs) [inline gradio/interface.py:182]: Will this work if someone passes in a `gr.Variable()` instead of `"state`"

File: gradio/routes.py [hunk 9]
```diff
@@ -308,15 +320,13 @@ def get_types(cls_set: List[Type], component: str):
...
+        " component. Please see the getting started for more information."
     )
 
 
 def set_state(*args):
     raise DeprecationWarning(
-        "This function is deprecated. To create stateful demos, pass 'state'"
-        "as both an input and output component. Please see the getting started"
-        "guide for more information."
+        "This function is deprecated. To create stateful demos, use the Variable"
```
Comment (abidlabs) [inline gradio/routes.py:330]: I think we should leave the old warnings. If someone is using `gr.set_state()` and `gr.get_state()`, they are probably trying to create an `Interface`, not a general `Blocks`

**Comment by abidlabs** [comment]
Left some minor comments, but overall LGTM.

File: gradio/interface.py (hunk 8 above)
Comment (aliabid94) [reply gradio/interface.py:182]: If they pass the same variable, yes. Meaning
```
var = gr.Variable()
gr.Interface(fn, ["text", var], ["text", var])
```
will work, but
```
gr.Interface(fn, ["text", gr.Variable()], ["text", gr.Variable()])
```
will not map to the same variable. Using the string "state" automatically creates the same Variable for both input and output.

File: demo/simple_squares/run.py (hunk 7 above)
Comment (aliabid94) [reply demo/simple_squares/run.py:1]: Why are we naming them like this? I'll change for now to be consistent but not a fan of this naming.

Comment (FarukOzderim) [reply demo/simple_squares/run.py]: I think one of the main advantages would be promoting and showing examples for the Blocks release.

File: gradio/blocks.py (hunk 4 above)
Comment (FarukOzderim) [reply gradio/blocks.py:257]: As far as I can see the else comes after the for loop and it enters to the `else` scope every time.
```
for loops also have an else clause which most of us are unfamiliar with. The else clause executes after the loop completes normally.
```
